sourcefinder/config.py

```diff
@@ -0,0 +1,137 @@
+from dataclasses import asdict
+from dataclasses import astuple
+from dataclasses import dataclass
+from dataclasses import field
+from dataclasses import is_dataclass
+from pathlib import Path
+import tomllib
+from types import UnionType
+from typing import get_args
+from typing import get_origin
+from typing import get_type_hints
+from typing import Mapping
+from typing import Container
+from typing import Type
+from typing import TypeVar
+
+T = TypeVar("T")
+
+
+def unguarded_is_dataclass(_type: Type[T], /) -> bool:
+    """Remove :ref:`TypeGuard` from is_dataclass.
+
+    see: https://github.com/python/mypy/issues/14941
+    """
+    return is_dataclass(_type)
+
+
+def validate_nested(value, type_):
+    origin_t = get_origin(type_)
+    args = get_args(origin_t)
+    if issubclass(origin_t, list):
+        assert isinstance(value, list), f"type({value}) != list"
+        if args:
+            for i in value:
+                validate_nested(i, args[0])
+    elif issubclass(origin_t, tuple):
+        assert isinstance(value, tuple), f"type({value}) != tuple"
+        for i, t in zip(value, args):
+            assert isinstance(i, t), f"type({i}) != {t}"
+    elif issubclass(origin_t, dict):
+        assert isinstance(value, dict)
+        for i in value.values():
+            assert isinstance(i, args[1]), f"type({i}) != {args[1]}"
+    else:
+        pass
+
+
+@dataclass(frozen=True)
+class _Validate:
+    def __post_init__(self):
+        for (key, type_), val in zip(get_type_hints(self).items(), astuple(self)):
+            match get_origin(type_):
+                case type() as origin_t:
+                    if issubclass(origin_t, Container):
+                        validate_nested(val, type_)
+                    elif issubclass(origin_t, UnionType):
+                        args = get_args(type_)
+                        assert isinstance(val, args), f"type({val}) != {args}"
+                    else:
+                        pass
+                case None:
+                    assert isinstance(val, type_), f"type({val}) != {type_}"
+
+
+_structuring_element = [[1, 1, 1], [1, 1, 1], [1, 1, 1]]
+_source_params = [
+    "ra",
+    "dec",
+    "peak",
+    "flux",
+    "sig",
+    "smaj_asec",
+    "smin_asec",
+    "theta_celes",
+    "ew_sys_err",
+    "ns_sys_err",
+    "error_radius",
+    "gaussian",
+    "chisq",
+    "reduced_chisq",
+]
+
+
+@dataclass(frozen=True)
+class ImgConf(_Validate):
+    interpolate_order: int = 1
+    median_filter: int = 0
+    mf_threshold: int = 0
+    rms_filter: float = 0.001
+    deblend_mincont: float = 0.005
+    structuring_element: list[list[int]] = field(
+        default_factory=lambda: _structuring_element
+    )
+    vectorized: bool = False
+    sep: int | None = None
```

Comment:
What does `float | None = None` establish?

Again, `sep` can be removed and should no longer be a dependency.

Reply:
The `= None` means that's the default. Since I didn't know what are appropriate values, I used `None`. I also guessed the types based on the names, so I probably got a lot of them wrong.